Ironware Support - Initial Parsers

jamesditrapani opened the pull request:
## Description
Initial PR to include Parsers for the Ironware NOS. Unicon support has been added via [PR](https://github.com/CiscoTestAutomation/unicon.plugins/pull/8).

Open to feedback/review of the included parsers.

**Includes the following parsers:**
* show interfaces brief
* show ip interfaces
* show media <interface>
* show mpls lsp
* show mpls vll <vll>
* show mpls vll-local <vll>
* show mpls ldp neighbor
* show optic <slot>
* show ip ospf neighbor
* show ip ospf interface brief
* show ip route
* show ip route summary

**Folder Based Tests have been included for all parsers.**

**All testing performed on the below devices:**
- Brocade MLX-8 - Ironware Version 5.8.0
- Brocade CES - Ironware Version 6.0.0

**PEP8 Compliant (Below errors are just comments)**
```
(pyats-env) [james.ditrapani@utils1][~/project_folder/pyats-dev/genieparser/src/genie/libs/parser/ironware]
$ pycodestyle *
show_interface.py:6:80: E501 line too long (81 > 79 characters)
show_media.py:6:80: E501 line too long (81 > 79 characters)
show_mpls.py:6:80: E501 line too long (81 > 79 characters)
show_optic.py:6:80: E501 line too long (81 > 79 characters)
show_ospf.py:6:80: E501 line too long (81 > 79 characters)
show_routing.py:6:80: E501 line too long (81 > 79 characters)
```

## Motivation and Context
Further NOS support for the PyATS Framework.

## Impact (If any)
N/A. These changes are only relevant to the Ironware NOS which doesn't exist in the Framework until now.

## Screenshots:
```
pyats-env) [james.ditrapani@utils1][~/project_folder/pyats-dev/genieparser/tests]
$ python ci_folder_parsing.py -o ironware
2021-01-05T21:24:53: %AETEST-INFO: +------------------------------------------------------------------------------+
2021-01-05T21:24:53: %AETEST-INFO: |                               Detailed Results                               |
2021-01-05T21:24:53: %AETEST-INFO: +------------------------------------------------------------------------------+
2021-01-05T21:24:53: %AETEST-INFO:  SECTIONS/TESTCASES                                                      RESULT
2021-01-05T21:24:53: %AETEST-INFO: --------------------------------------------------------------------------------
2021-01-05T21:24:53: %AETEST-INFO: .
2021-01-05T21:24:53: %AETEST-INFO: `-- FileBasedTest                                                         PASSED
2021-01-05T21:24:53: %AETEST-INFO:     |-- setup                                                             PASSED
2021-01-05T21:24:53: %AETEST-INFO:     `-- test[operating_system=ironware]                                   PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 1: ironware -> ShowMediaInterface                        PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 1.1: Test Golden -> ironware -> ShowMediaInterface       PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 1.1.1: Gold -> ironware -> ShowMediaInterface -> c...    PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 1.1.2: Gold -> ironware -> ShowMediaInterface -> f...    PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 1.1.3: Gold -> ironware -> ShowMediaInterface -> n...    PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 1.2: Test Empty -> ironware -> ShowMediaInterface        PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 1.2.1: Empty -> ironware -> ShowMediaInterface -> ...    PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 2: ironware -> ShowIPOSPFInterfaceBrief                  PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 2.1: Test Golden -> ironware -> ShowIPOSPFInterfac...    PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 2.1.1: Gold -> ironware -> ShowIPOSPFInterfaceBrie...    PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 2.1.2: Gold -> ironware -> ShowIPOSPFInterfaceBrie...    PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 2.2: Test Empty -> ironware -> ShowIPOSPFInterface...    PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 2.2.1: Empty -> ironware -> ShowIPOSPFInterfaceBri...    PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 3: ironware -> ShowIPOSPFNeighbor                        PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 3.1: Test Golden -> ironware -> ShowIPOSPFNeighbor       PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 3.1.1: Gold -> ironware -> ShowIPOSPFNeighbor -> d...    PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 3.1.2: Gold -> ironware -> ShowIPOSPFNeighbor -> l...    PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 3.2: Test Empty -> ironware -> ShowIPOSPFNeighbor        PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 3.2.1: Empty -> ironware -> ShowIPOSPFNeighbor -> ...    PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 4: ironware -> ShowMPLSLDPNeighbor                       PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 4.1: Test Golden -> ironware -> ShowMPLSLDPNeighbor      PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 4.1.1: Gold -> ironware -> ShowMPLSLDPNeighbor -> ...    PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 4.2: Test Empty -> ironware -> ShowMPLSLDPNeighbor       PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 4.2.1: Empty -> ironware -> ShowMPLSLDPNeighbor ->...    PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 5: ironware -> ShowMPLSLSP                               PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 5.1: Test Golden -> ironware -> ShowMPLSLSP              PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 5.1.1: Gold -> ironware -> ShowMPLSLSP -> golden_o...    PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 5.2: Test Empty -> ironware -> ShowMPLSLSP               PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 5.2.1: Empty -> ironware -> ShowMPLSLSP -> empty_o...    PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 6: ironware -> ShowMPLSVLL                               PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 6.1: Test Golden -> ironware -> ShowMPLSVLL              PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 6.1.1: Gold -> ironware -> ShowMPLSVLL -> down_output    PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 6.1.2: Gold -> ironware -> ShowMPLSVLL -> normal_o...    PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 6.1.3: Gold -> ironware -> ShowMPLSVLL -> qinq_output    PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 6.1.4: Gold -> ironware -> ShowMPLSVLL -> untagged...    PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 6.2: Test Empty -> ironware -> ShowMPLSVLL               PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 6.2.1: Empty -> ironware -> ShowMPLSVLL -> empty_o...    PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 7: ironware -> ShowMPLSVLLLocal                          PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 7.1: Test Golden -> ironware -> ShowMPLSVLLLocal         PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 7.1.1: Gold -> ironware -> ShowMPLSVLLLocal -> nor...    PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 7.1.2: Gold -> ironware -> ShowMPLSVLLLocal -> qin...    PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 7.1.3: Gold -> ironware -> ShowMPLSVLLLocal -> unt...    PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 7.2: Test Empty -> ironware -> ShowMPLSVLLLocal          PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 7.2.1: Empty -> ironware -> ShowMPLSVLLLocal -> em...    PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 8: ironware -> ShowIPInterface                           PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 8.1: Test Golden -> ironware -> ShowIPInterface          PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 8.1.1: Gold -> ironware -> ShowIPInterface -> gold...    PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 8.2: Test Empty -> ironware -> ShowIPInterface           PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 8.2.1: Empty -> ironware -> ShowIPInterface -> emp...    PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 9: ironware -> ShowInterfacesBrief                       PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 9.1: Test Golden -> ironware -> ShowInterfacesBrief      PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 9.1.1: Gold -> ironware -> ShowInterfacesBrief -> ...    PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 9.2: Test Empty -> ironware -> ShowInterfacesBrief       PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 9.2.1: Empty -> ironware -> ShowInterfacesBrief ->...    PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 10: ironware -> ShowIPRoute                              PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 10.1: Test Golden -> ironware -> ShowIPRoute             PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 10.1.1: Gold -> ironware -> ShowIPRoute -> golden_...    PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 10.2: Test Empty -> ironware -> ShowIPRoute              PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 10.2.1: Empty -> ironware -> ShowIPRoute -> empty_...    PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 11: ironware -> ShowIPRouteSummary                       PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 11.1: Test Golden -> ironware -> ShowIPRouteSummary      PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 11.1.1: Gold -> ironware -> ShowIPRouteSummary -> ...    PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 11.2: Test Empty -> ironware -> ShowIPRouteSummary       PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 11.2.1: Empty -> ironware -> ShowIPRouteSummary ->...    PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 12: ironware -> ShowOptic                                PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 12.1: Test Golden -> ironware -> ShowOptic               PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 12.1.1: Gold -> ironware -> ShowOptic -> new_optic...    PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 12.1.2: Gold -> ironware -> ShowOptic -> no_optic_...    PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 12.1.3: Gold -> ironware -> ShowOptic -> normal_ou...    PASSED
2021-01-05T21:24:53: %AETEST-INFO:         |-- Step 12.2: Test Empty -> ironware -> ShowOptic                PASSED
2021-01-05T21:24:53: %AETEST-INFO:         `-- Step 12.2.1: Empty -> ironware -> ShowOptic -> empty_ou...    PASSED
2021-01-05T21:24:53: %AETEST-INFO: +------------------------------------------------------------------------------+
2021-01-05T21:24:53: %AETEST-INFO: |                                   Summary                                    |
2021-01-05T21:24:53: %AETEST-INFO: +------------------------------------------------------------------------------+
2021-01-05T21:24:53: %AETEST-INFO:  Number of ABORTED                                                            0
2021-01-05T21:24:53: %AETEST-INFO:  Number of BLOCKED                                                            0
2021-01-05T21:24:53: %AETEST-INFO:  Number of ERRORED                                                            0
2021-01-05T21:24:53: %AETEST-INFO:  Number of FAILED                                                             0
2021-01-05T21:24:53: %AETEST-INFO:  Number of PASSED                                                             1
2021-01-05T21:24:53: %AETEST-INFO:  Number of PASSX                                                              0
2021-01-05T21:24:53: %AETEST-INFO:  Number of SKIPPED                                                            0
2021-01-05T21:24:53: %AETEST-INFO:  Total Number                                                                 1
2021-01-05T21:24:53: %AETEST-INFO:  Success Rate                                                            100.0%
2021-01-05T21:24:53: %AETEST-INFO: --------------------------------------------------------------------------------
```

## Checklist:
<!--- This is meant more as a personal checklist so we don't forgot important steps! -->
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] I have updated the changelog.
- [ ] I have updated the documentation (If applicable).
- [x] I have added tests to cover my changes (If applicable).
- [x] All new and existing tests passed.
- [x] All new code passed compilation.

KamyarZiabari commented on src/genie/libs/parser/ironware/show_optic.py:
```diff
@@ -0,0 +1,153 @@
+                rx = m.groupdict()['rx']
+                tx = m.groupdict()['tx']
+                tbc = m.groupdict()['tbc']
+
```
Just a note, you can do this as well and update the dict at once. I think this is trimmer personally, and I suggest that you apply this way, but I leave it to you to decide, you can try it and if you find it appropriate, apply it here and elsewhere
```
interfaces_dict.update({port: {
                                         'temperature': {'alarm': temp},
                                          'rx': {'alarm': rx},
                                          'tx': {'alarm': tx},
                                          'tx_bias_current': {'alarm': tbc},

}
})

jamesditrapani replied:
@KamyarZiabari It is my understanding that using `interfaces_dict.update` will actually update the inner keys for `temperature`, `rx`, `tx` and `tx_bias_current` with purely the subdict containing `{'alarm': xxx}`. E.g:

```python
>>> mydict1 = {
...     'interfaces': {
...             'temp': {
...                     'value': 1234
...             }
...     }
... }
>>>
>>> print(mydict1)
{'interfaces': {'temp': {'value': 1234}}}
>>>
>>>
>>> mydict1.update({'interfaces': {
...     'temp': {'alarm': True}
...     }
... }
... )
>>> print(mydict1)
{'interfaces': {'temp': {'alarm': True}}}
>>>
```

KamyarZiabari commented on src/genie/libs/parser/ironware/show_ospf.py:
```diff
@@ -0,0 +1,207 @@
+
+                group = m.groupdict()
+                neighbor = group['neighbor']
+                neigh_dict = neighbors.setdefault(neighbor, {
```
why setdefault and not update

KamyarZiabari commented on src/genie/libs/parser/ironware/show_routing.py:
```diff
@@ -0,0 +1,307 @@
+                if 'routes' not in ip_dict:
+                    result_dict = ip_dict.setdefault('routes', {})
+
+                result_dict['{0}/{1}'.format(network, cidr)] = {
```
Nice!

KamyarZiabari commented:
AMAZING work, thanks alot for contributing to pyATS. 
I added my comments. 
@tahigash will review the schemas, and when comments addressed. it would be ready to merge.

Thanks again for your contribution.

jamesditrapani commented on src/genie/libs/parser/ironware/show_mpls.py:
```diff
@@ -0,0 +1,691 @@
+
+
+# ====================================================
+#  parser for 'show mpls lsp wide'
```
@KamyarZiabari Brocade/Ironware has no concept of `terminal width 0` like Cisco. As such, wide needs to be appended to commands to prevent them being compressed. Here is an example:

`show mpls lsp`:
```
ssh@edge1.code1.city#sh mpls lsp
Note: LSPs marked with * are taking a Secondary Path
                               Admin Oper  Tunnel   Up/Dn Retry Active
Name           To              State State Intf     Times No.   Path
edge1.code1   192.168.1.2  UP    UP    tnl3     2     0     --
.city_to_lsr1
.code2.city
edge1.code1   192.168.1.137  UP    UP    tnl0     2     0     --
.city_to_lsr3
.code3.city
edge1.code1   192.168.1.24   UP    UP    tnl2     2     0     --
.city_to_lsr3
.code4.city
edge1.code1   192.168.1.138  UP    UP    tnl1     1     0     --
.city_to_lsr5
.code5.city
```

`show mpls lsp wide`:
```
ssh@edge1.code1.city#sh mpls lsp wide
Note: LSPs marked with * are taking a Secondary Path
                                                                                 Admin Oper  Tunnel   Up/Dn Retry Active
Name                                                             To              State State Intf     Times No.   Path
edge1.code1.city_to_lsr1.code2.city                              192.168.1.2  UP    UP    tnl3     2     0     --                    
edge1.code1.city_to_lsr3.code3.city                              192.168.1.137  UP    UP    tnl0     2     0     --                    
edge1.code1.city_to_lsr3.code4.city                              192.168.1.24   UP    UP    tnl2     2     0     --                    
edge1.code1.city_to_lsr5.code5.city                              192.168.1.138  UP    UP    tnl1     1     0     --
```

Let me know if this is okay.

tahigash commented on src/genie/libs/parser/ironware/tests/ShowIPInterface/cli/equal/golden_output1_expected.py:
```diff
@@ -0,0 +1,53 @@
+expected_output = {
```
schema looks good.

tahigash commented on src/genie/libs/parser/ironware/tests/ShowIPOSPFInterfaceBrief/cli/equal/documentation_output_expected.py:
```diff
@@ -0,0 +1,13 @@
+expected_output = {
```
schema looks good

tahigash commented on src/genie/libs/parser/ironware/tests/ShowIPOSPFInterfaceBrief/cli/equal/documentation_output_expected.py:
```diff
@@ -0,0 +1,13 @@
+expected_output = {
+  'total': 1,
+  'interfaces': {
+    'ethernet1/2': {
```
converting from `eth 1/2` to `ethernet1/2`?

jamesditrapani replied:
@tahigash Correct. Expanding interface names in the parsers. Is this non standard?

tahigash replied:
this is ok. we are doing the same on XE/XR/NXOS. just wanted to confirm.

tahigash commented on src/genie/libs/parser/ironware/tests/ShowIPOSPFInterfaceBrief/cli/equal/documentation_output_expected.py:
```diff
@@ -0,0 +1,13 @@
+expected_output = {
```
for these files, under `equal` folder, we usually have like below file name with prefix `golden_`.

```
golden_output_expected.py
golden_output2_expected.py

jamesditrapani replied:
@tahigash I see, are you happy to leave the files named like this? Gives a bit more description about what they're testing against. E.g. documentation_x_x is testing against output displayed in the Ironware Documentation whereas live_x_x is testing against a devices live output.

tahigash replied:
ok. we can keep current names. having meaningful name would be more useful.

tahigash commented on src/genie/libs/parser/ironware/tests/ShowIPOSPFNeighbor/cli/equal/documentation_output_expected.py:
```diff
@@ -0,0 +1,54 @@
+expected_output = {
```
this device supports like `vrf`? if yes, I assume this output belongs to `default` vrf.  for routing protocols, we usually have `vrf` level like below:
```
'vrf': {
    'default': {
        'neighbors': {
            '10.1.10.2': {
(snip)
```

jamesditrapani replied:
@tahigash Device does support VRFs however to display OSPF neighbours in a VRF the command is `show ip ospf vrf xxx neighbor`, I'm still happy to display this as the default if you wish?

tahigash replied:
yes. please add `vrf` level to the parser. otherwise, when someone want to add `show ip ospf vrf xxx neighbor` command, it will need to have different schema. better to leverage this one.

tahigash commented on src/genie/libs/parser/ironware/tests/ShowIPRoute/cli/equal/golden_output1_expected.py:
```diff
@@ -0,0 +1,188 @@
+  'routes': {
+    '10.200.0.12/30': {
+      'network': '10.200.0.12',
+      'cidr': 30,
```
`cidr` -> `netmask` is better?

tahigash replied:
confirmed the change

tahigash commented on src/genie/libs/parser/ironware/tests/ShowIPRoute/cli/equal/golden_output1_expected.py:
```diff
@@ -0,0 +1,188 @@
+expected_output = {
```
this is `default` vrf routing table? like other comment, would suggest to have `vrf` top level.

jamesditrapani replied:
@tahigash Yes, similar to above however the command is `show ip route vrf xx`, it can be assumed when running `show ip route` that the user is looking at the default table.

tahigash replied:
ok. if so, can you add `vrf` level like ospf parser? it's same reason with the ospf parser.

tahigash commented on src/genie/libs/parser/ironware/tests/ShowIPRouteSummary/cli/equal/golden_output_output.txt:
```diff
@@ -0,0 +1,5 @@
+  3 connected, 0 static, 0 RIP, 446 OSPF, 0 BGP, 0 ISIS
+  Number of prefixes:
+  /24: 1 /26: 2 /27: 3 /28: 5 /29: 1 /30: 34 /31: 242 /32: 161
+Nexthop Table Entry - 5 entries
```
this line was not parsed?

jamesditrapani replied:
Fixed

tahigash commented on src/genie/libs/parser/ironware/tests/ShowIPRouteSummary/cli/equal/golden_output_expected.py:
```diff
@@ -0,0 +1,21 @@
+    'bgp': 0,
+    'isis': 0
+  },
+  'cidrs': {
```
I understand `cidrs`, but might be better to have similar word with output for everyone.
```
`number_of_prefixes`: {
    '/24': 1,
(snip)

jamesditrapani replied:
I will change this to `netmask` too.

tahigash replied:
confirmed the change.

tahigash commented on src/genie/libs/parser/ironware/tests/ShowMPLSLDPNeighbor/cli/equal/documentation_output_expected.py:
```diff
@@ -0,0 +1,22 @@
+expected_output = {
```
schema looks good

tahigash commented on src/genie/libs/parser/ironware/tests/ShowMPLSLSP/cli/equal/golden_output1_expected.py:
```diff
@@ -0,0 +1,35 @@
+            'operational': 'UP',
+            'flap_count': 1,
+            'retry_count': 0,
+            'tunnel_interface': 'tnl63'
```
in some parser, short interface name converted to full interface name. I would suggest to do the same to all parsers. so that we can see full interface name all the time and easy to match/compare between parser outputs later.

jamesditrapani replied:
Noted, have fixed up

tahigash commented on src/genie/libs/parser/ironware/tests/ShowMPLSLSP/cli/equal/golden_output1_expected.py:
```diff
@@ -0,0 +1,35 @@
+expected_output = {
```
schema looks good

tahigash commented on src/genie/libs/parser/ironware/tests/ShowMPLSVLL/cli/equal/normal_output_expected.py:
```diff
@@ -0,0 +1,36 @@
+        'extended_counters': True,
+        'counters': False
+      },
+      'peer': {
```
if you might have multiple peers, you might want to have like below:
```
'peer': {
    '192.168.1.1': {
        'state': 'UP',

jamesditrapani replied:
@tahigash A VLL is a point-to-point L2VPN pseudowire, it will never have more than one peer. Going to keep as is for now.

tahigash replied:
ok. if so, no issue. thanks for the confirmation!

tahigash commented on src/genie/libs/parser/ironware/tests/ShowMPLSVLLLocal/cli/equal/normal_output_expected.py:
```diff
@@ -0,0 +1,22 @@
+      'vll_id': 3,
+      'ifl_id': '--',
+      'state': 'UP',
+      'endpoint_1': {
```
below might be better:
```
'endpoint': {
    1: {
        'type': 'tagged',

jamesditrapani replied:
Agree, changed Schema to reflect this.

tahigash reviewed (commented):
thank you for your contribution! I commented some for schema. but no big issue. if you want to keep current way, we can keep. it's open-source. so if any issue, we can update at that point.

jamesditrapani commented:
> thank you for your contribution! I commented some for schema. but no big issue. if you want to keep current way, we can keep. it's open-source. so if any issue, we can update at that point.

Thanks @tahigash, I've pushed another commit with your suggested changes.

tahigash reviewed (changes requested):
@jamesditrapani thanks for addressing comments. as I added comments, please add `vrf` level to route/ospf parsers.

jamesditrapani commented:
> @jamesditrapani thanks for addressing comments. as I added comments, please add `vrf` level to route/ospf parsers.

Thanks Takashi, I've updated these Schemas now.

tahigash reviewed (approved):
thanks for addressing my comments. schema looks good to me now.

@KamyarZiabari will finish review for codes.

KamyarZiabari commented on src/genie/libs/parser/ironware/show_mpls.py:
```diff
@@ -0,0 +1,691 @@
+                    'vll_index': int(m.groupdict()['vllindex'])
+                })
+                continue
+
```
@jamesditrapani  You havent addressed this comment.
It is necessary to have these comments since it is large parser. When re-reading it would help us see the line example right there. 
So please add a comment on top of all line similar to below
```
m = p2.match(line)

jamesditrapani replied:
@KamyarZiabari Sorry, looks like I missed this one originally!

KamyarZiabari commented on src/genie/libs/parser/ironware/show_mpls.py:
```diff
@@ -0,0 +1,709 @@
+                            m.groupdict()['state'] is not None else 'Unknown'
+                continue
+
+            m = p3.match(line)
```
Same comment as above please add the line example:
e.g :
```
 #Counter          : disabled
m = p6.match(line)

KamyarZiabari commented on src/genie/libs/parser/ironware/show_optic.py:
```diff
@@ -0,0 +1,153 @@
+
+        result_dict = {}
+
+        p1 = re.compile(r'(^(?P<port>\d+\/\d+)\s+'
```
add line that matches the pattern as comment similar to other mdoules

jamesditrapani commented:
@KamyarZiabari Cheers, fixed those up now, looks like I just missed a few :)

What would you like me to do about the `github_parser.json` merge conflict?

KamyarZiabari commented:
> @KamyarZiabari Cheers, fixed those up now, looks like I just missed a few :)
> 
> What would you like me to do about the `github_parser.json` merge conflict?

Can you please otherwise you cant merge it. I cannot see the conflicts on github

tahigash commented:
@jamesditrapani thanks for contribution again! merged!
